Close overlay menu when clicking an anchor link

Trapsta opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR adds an event Listener to toggle the navigation block overlay menu on clicking links inside the overlay.

Fixes #39585

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Navigating to anchor links in the navigation block in an overlay menu does not automatically close the overlay. This requires manually closing the modal to view the link's content.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Since we are using [Micromodal](https://github.com/ghosh/Micromodal) to implement the overlay menu, am using the `MicroModal.close()` method to toggle the modal on clicking the navigation link. My initial idea was to use the `data-micromodal-close` custom attribute which Micromodal uses as a trigger to close the modal but I couldn't get it to work reliably. The overlay would close but sometimes the browser would not navigate to the correct anchor.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Open a post or page.
2. Create a navigation menu and add an anchor link.
3. View the post/page on the frontend.
4. Make your browser screen smaller, below 600px.
5. Open the menu and click the anchor link.
7.  Observe that the overlay menu modal closes and page navigates to the anchor link's content. Try also testing by changing the navigation block overlay menu display  setting to `off` , `always` or `mobile`.


## Screenshots or screencast <!-- if applicable -->

#### Before
https://user-images.githubusercontent.com/276316/159422355-a6ad6d65-e866-4874-94fa-9fa6f4beca87.mov 

### After
https://user-images.githubusercontent.com/276316/159422443-e9817778-5499-4590-bd9d-f39abf81040c.mov


Fixes #39585

Trapsta committed:
fix(block-library): close overlay menu in nav block when navigating to an anchor link

Closes #gutenberg-39585

Trapsta committed:
fix(block-library): add eventlistener to close modal on clicking overlay menu nav links

Closes #gutenberg-39585

getdave commented:
Thanks for the PR! 👏 

I'm wondering could you complete the `How` section to describe why you feel this is the best approach and what alternatives you may have considered.

I was considering tackling this so it's great to see it's already being addressed. Thank you!

JosVelasco commented:
Hello @Trapsta. Thanks for the PR. For testing purposes consider that you can also set the overlay menu to be visible always.

Collapses the navigation options in a menu icon opening an overlay:
Off | Mobile | Always

Trapsta commented:
Thanks @getdave , sure thing - I will add some context on that. Noted @JosVelasco , thanks for the heads up , I had probably overlooked that part

getdave commented:
> Thanks @getdave , sure thing - I will add some context on that. Noted @JosVelasco , thanks for the heads up , I had probably overlooked that part

Thanks. Let me know when it's ready for review.

Trapsta committed:
fix(block-library): pass id of modal to micromodal.close and check if overlay menu is open before trying to close it

Closes #gutenberg-39585

Trapsta commented:
@getdave @JosVelasco  I have updated the PR, please review when you can

paaljoachim commented:
I was asked by Jos to take a look at this PR.
I am following this method: https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/

Testing.
Local site 
WP 6.0 Beta 1.
Twenty Twenty One
Chrome
OSX on my Macbook Pro.

Opened an existing page.
Added a Navigation block.
Added a few random links to pages and posts.
Added an anchor link.
Noticed that the Navigation block also used the dot save system. Saved the page and the Navigation Menus.

Previewed on the frontend.
Opened up Chrome dev tools and made the screen smaller than 600px.
Noticed the menu is now showing the hamburger/mobile menu.
Clicked the anchor link.

Overlay menu modal closes and the page navigates to the anchor link's content.

Backend Navigation block settings.
Overlay Menu: Off - Mobile - Always. **All three worked nicely with this PR fix.**

Thank you @Trapsta !

Ps.
**We should probably test this in multiple browsers.....**

paaljoachim commented:
As this is a bug fix I went ahead and added the backport to beta label.

paaljoachim reviewed (commented):
My focus was on testing the functionality that it worked as it should using Chrome. 
The fix works very well!

JosVelasco commented:
Also tested the PR on my end successfully. I even installed the popular plugin Page scroll to id to animate the scroll and it worked like a charm. Thanks @Trapsta.

Tested on
Local site
WordPress 5.9.3 and then WordPress 6.0-beta1 
Twenty Twenty Two
Latest Firefox, Chrome and Safari.
OSX on my Macbook Air.

https://www.youtube.com/watch?v=uyP1rrljjuE

getdave commented on packages/block-library/src/navigation/view.js:
```diff
@@ -79,6 +79,30 @@ window.addEventListener( 'load', () => {
+		'.wp-block-navigation-item__content'
+	);
+
+	navigationLinks.forEach( function ( link ) {
+		// we need to find the specific parent modal for this link
+		// since .close() won't work without an ID in case we have
+		// mutiple navigation menus in a post/page.
+		const modal = link.closest(
```
```suggestion
	navigationLinks.forEach( function ( link ) {
	
		// Ignore non-anchor links.
		if ( ! link.getAttribute( 'href' )?.startsWith( '#' ) ) {
			return;
		}
		
		// we need to find the specific parent modal for this link
		// since .close() won't work without an ID in case we have
		// mutiple navigation menus in a post/page.
		const modal = link.closest(
```

We can ignore links that are no anchors.

draganescu replied:
This would also need to update the comment 

// Close modal on clicking internal and external links inside modal.

scruffian replied:
I think this makes sense. The other cases I thought about where:

- What if the link opens in new tab?
- What if the link is slow to load or is broken? 

But I think in both of these cases we wouldn't want to close the overlay.

getdave reviewed (commented):
Thank you for this PR. I tested it and it worked as expected.


https://user-images.githubusercontent.com/444434/165544578-17bd9fbd-33a7-4392-ae9f-e79c6f9142fc.mp4


Once issue I noticed was that for non-anchor links, the modal will close prior to Navigation. That might not always be desirable and we can guard against it by testing for an anchor before we add the handlers.

Trapsta committed:
fix(block-library): ignore non-anchor links when closing navigation menu overlay for anchor links

Closes #gutenberg-39585

Trapsta committed:
chore(block-library): move code for closing modal on clicking anchor links to seperate view-modal.js script

Closes #gutenberg-39585

Trapsta committed:
fix(block-library): ignore anchor links which open on a new tab when closing modal on clicking link

Closes #gutenberg-39585

Trapsta commented:
Great feedback , I can see adding the filter @getdave suggested enhances the UX for non-anchor links quite a bit and I also added a filter for any anchor links that would open in a new tab to cover some the edge cases @scruffian pointed out.  Also updated the comment to be more accurate. Thanks everyone for taking your time to test and review.

getdave commented on packages/block-library/src/navigation/view-modal.js:
```diff
@@ -33,4 +33,36 @@ window.addEventListener( 'load', () => {
+			return;
+		}
+
+		// we need to find the specific parent modal for this link
+		// since .close() won't work without an ID in case we have
+		// mutiple navigation menus in a post/page.
```
```suggestion
		// Find the specific parent modal for this link
		// since .close() won't work without an ID if there are
		// multiple navigation menus in a post/page.
```

getdave commented on packages/block-library/src/navigation/view-modal.js:
```diff
@@ -33,4 +33,36 @@ window.addEventListener( 'load', () => {
+		link.addEventListener( 'click', () => {
+			// check if modal exists and is open before we try to close it
+			// otherwise Micromodal will toggle the `has-modal-open` class
+			// on the html tag which prevents scrolling
```
```suggestion
			// check if modal exists and is open before trying to close it
			// otherwise Micromodal will toggle the `has-modal-open` class
			// on the html tag which prevents scrolling
```

getdave reviewed (approved):
This looks good to go. I've also pinged @scruffian and @draganescu for their opinions but I'd say this is a good improvement. Thank you again for your work here 🙇

Trapsta committed:
Update packages/block-library/src/navigation/view-modal.js

update comment description

Co-authored-by: Dave Smith <getdavemail@gmail.com>

Trapsta committed:
Update packages/block-library/src/navigation/view-modal.js

fix comment typo

Co-authored-by: Dave Smith <getdavemail@gmail.com>

aristath reviewed (approved):
LGTM! 👍

getdave commented:
Thank you again @Trapsta for the contribution.

adamziel commented:
I am removing the Backport label here. This PR depends on https://github.com/WordPress/gutenberg/pull/36176 which wasn't backported in time for 6.0 and is now planned for 6.1.

coreyworrell commented:
If this was the fix recently included in 6.1, it does not cover the common case of a link like `/#anchor` or `/some-page/#anchor`. It only works if the `href` exactly starts with `#`. So if you are on `/some-page/` and you click the `/some-page/#anchor` link, it will scroll down to the anchor, but the modal will remain open.

My current fix:

```javascript
import MicroModal from 'micromodal'

// Fix navigation block
const links = document.querySelectorAll('.wp-block-navigation-item__content')

for (const link of links) {
    const modal = link.closest('.wp-block-navigation__responsive-container')
    const id = modal?.id || ''

    if (id && link.href.indexOf('#') !== -1 && link.pathname === location.pathname) {
        link.addEventListener('click', () => {
            if (modal.classList.contains('has-modal-open')) {
                MicroModal.close(id)
            }
        })
    }
}
```

getdave commented:
> If this was the fix recently included in 6.1, it does not cover the common case of a link like `/#anchor` or `/some-page/#anchor`. It only works if the `href` exactly starts with `#`. So if you are on `/some-page/` and you click the `/some-page/#anchor` link, it will scroll down to the anchor, but the modal will remain open.
> 
> My current fix:
> 
> ```js
> import MicroModal from 'micromodal'
> 
> // Fix navigation block
> const links = document.querySelectorAll('.wp-block-navigation-item__content')
> 
> for (const link of links) {
>     const modal = link.closest('.wp-block-navigation__responsive-container')
>     const id = modal?.id || ''
> 
>     if (id && link.href.indexOf('#') !== -1 && link.pathname === location.pathname) {
>         link.addEventListener('click', () => {
>             if (modal.classList.contains('has-modal-open')) {
>                 MicroModal.close(id)
>             }
>         })
>     }
> }
> ```

@draganescu I looks like we'll need a bug fix PR for this.
